Yoast/wordpress-seo
# Add WP-CLI commands to manage redirects

## Summary

This PR can be summarized in the following changelog entry:

* Adds WP-CLI commands to manage redirects.

## Relevant technical choices:

* The following subcommands are implemented under the `redirect` command root:
    `redirect create` - Creates a new Yoast SEO redirect.
    `redirect delete` - Deletes an existing Yoast SEO redirect.
    `redirect follow` - Follows a Yoast SEO redirect chain to get the final target it resolves to.
    `redirect has` - Checks whether a given Yoast SEO redirect exists.
    `redirect list` - Lists Yoast SEO redirects.
    `redirect update` - Updates an existing Yoast SEO redirect.
* The check for premium has been moved out of the individual commands and will bew processed on the `before_invoke` hook for each premium command. The code can be deployed with the free codebase, as the logic is not contained within the commands, and a meaningful error message is produced for premium commands run on the free version.
* The parameter to select the output rendering format does not use the convention set by the bundled WP-CLI commands (`--format=<format>`). Instead, I opted to keep the `--format` for easy filtering of the redirect "format", and moved the output rendering flag to `--output=<output>` instead.
* The commands are built for convenient scripting. Here are a few examples of how that works:
    ```
    # redirect has will return an exit code, so that it can be directly used in shell conditionals:
    $ URL="<some-url>"; if wp redirect has $URL; then
        echo "Yes, the redirect for $URL exists."
      else
        echo "No, the redirect for $URL does not exist."
      fi
    No, the redirect for <some-url> does not exist.

    # redirect list output can be filtered and shaped so that it can be fed into other commands.
    # As an example, here's how to delete all 307 redirects:
    wp redirect list --type=307 --field=origin | xargs -I {} wp redirect delete {}
* As unit tests are mostly worthless for commands (as the logic is found elsewhere), and as we don't have an integration test infrastructure yet, this pull request does not contain any automated tests. I suggest creating a new issue for tests and implement them as soon as an integration testing infrastructure is available.

## Test instructions

This PR can be tested by following these steps:

* Enter the shell terminal.
* Ensure you have installed a working version of WP-CLI by running `wp cli info` ([Installation instructions](https://make.wordpress.org/cli/handbook/installing/#recommended-installation)).
* Change into the root folder of the development site where you are running Yoast SEO with this pull request.
* Running `wp help redirect` should now show an overview of the available redirect subcommands.
* Running `wp help redirect <subcommand>` will produce a help screen for that given subcommand.
* Using a subcommand should be explained on the respective help screens.

## Quality assurance


Fixes https://github.com/Yoast/wordpress-seo-premium/issues/1566
